query-translator/src/translate.rs

```diff
@@ -427,14 +427,47 @@ fn re_project(mut inner: SelectQuery, projection: Projection) -> SelectQuery {
     let limit = inner.limit;
     inner.limit = Limit::None;
 
-    SelectQuery {
-        distinct: outer_distinct,
-        projection: projection,
-        from: FromClause::TableList(TableList(vec![TableOrSubquery::Subquery(Box::new(inner))])),
-        constraints: vec![],
-        group_by: group_by,
-        order: order_by,
-        limit: limit,
+    use self::Projection::*;
+
+    let nullable = match &projection {
+        &Columns(ref columns) => {
+            columns.iter().filter_map(|pc| {
+                match pc {
+                    &ProjectedColumn(ColumnOrExpression::NullableAggregate(_, _), ref name) => {
+                        Some(Constraint::IsNotNull {
+                            value: ColumnOrExpression::ExistingColumn(name.clone()),
+                        })
+                    },
+                    _ => None,
+                }
+            }).collect()
+        },
+        &Star => vec![],
+        &One => vec![],
+    };
+
+    let outer = SelectQuery {
+            distinct: outer_distinct,
+            projection: projection,
+            from: FromClause::TableList(TableList(vec![TableOrSubquery::Subquery(Box::new(inner))])),
+            constraints: vec![],
+            group_by: group_by,
+            order: order_by,
+            limit: limit,
+        };
+
+    if nullable.is_empty() {
+        outer
+    } else {
+        SelectQuery {
+            distinct: false,
+            projection: Projection::Star,
+            from: FromClause::TableList(TableList(vec![TableOrSubquery::Subquery(Box::new(outer))])),
+            constraints: nullable,
+            group_by: vec![],
+            order: vec![],
```

Comment:
Gosh. I don’t actually know if the inner ordering is respected when returned in this way; there are databases in which it is not. Worth a little research and a test…